### PR TITLE
UI controller upgrade support for Linux

### PR DESCRIPTION
**Pull request by foxriver76**

### CODE REVIEW

File: packages/controller/src/lib/upgradeManager.ts [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+        return { version, adminInstance };
+    }
+
+    /**
+     * Stops the js-controller via cli call
+     */
+    async stopController(): Promise<void> {
+        await execAsync(`${tools.appNameLowerCase} stop`);
+        await wait(STOP_TIMEOUT_MS);
```
Comment (Apollon77) [inline packages/controller/src/lib/upgradeManager.ts:101]: wait only 3s?

File: packages/controller/src/lib/upgradeManager.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+        const obj = await objects.getObjectAsync('system.certificates');
+
+        if (!obj) {
+            throw new Error('No certificates found');
+        }
+
+        const certs = obj.native.certificates;
+
+        return { certPrivate: certs[certPrivateName], certPublic: certs[certPublicName] };
```
Comment (Apollon77) [inline packages/controller/src/lib/upgradeManager.ts:223]: does that work with acme too? :-)

File: packages/controller/src/main.ts [hunk 3]
```diff
@@ -3132,6 +3129,40 @@ async function processMessage(msg: ioBroker.SendableMessage): Promise<null | voi
             break;
         }
 
+        case 'upgradeController': {
+            if (['win32', 'darwin'].includes(os.platform())) {
```
Comment (Apollon77) [inline packages/controller/src/main.ts:3133]: can we also detect "non-systemd variants?"

File: packages/controller/src/lib/upgradeManager.ts [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    }
+
+    /**
+     * This function is called when the webserver receives a message
+     *
+     * @param req received message
+     * @param res server response
+     * @param server https or https server instance
+     */
+    webServerCallback(req: http.IncomingMessage, res: http.ServerResponse, server: http.Server | https.Server): void {
```
Comment (Apollon77) [inline packages/controller/src/lib/upgradeManager.ts:157]: and this works to stream infos through?

**Review by Apollon77 — APPROVED** [review]
Hhmm ... ok ... I assume we need to merge it and then iterate with @GermanBluefox on it to finalize it for a good user experience, or?!

Additional Throught: The admin instance from host A (for port) does say nothing if the port is used on host B ... ... ... and also admin on hostA will not go offline if controller on host B does ...

File: packages/controller/src/lib/upgradeManager.ts (hunk 4 above)
Comment (foxriver76) [reply packages/controller/src/lib/upgradeManager.ts:157]: Correct.

File: packages/controller/src/lib/upgradeManager.ts (hunk 2 above)
Comment (foxriver76) [reply packages/controller/src/lib/upgradeManager.ts:223]: tbh: I hope so, the certificate names in the interal object should stay the same

File: packages/controller/src/lib/upgradeManager.ts (hunk 1 above)
Comment (foxriver76) [reply packages/controller/src/lib/upgradeManager.ts:101]: it was more than enough during my tests. also the stop command itself is awaited.

File: packages/controller/src/main.ts (hunk 3 above)
Comment (foxriver76) [reply packages/controller/src/main.ts:3132]: We don’t have this detection yet, if needed we can look into it, there sure will be a way. 

However, as we are calling `iob stop/start` it should then be handled there and not in this place if adaption is needed

**Pull request merged**